docs/development/building-good-containers.md

```diff
@@ -0,0 +1,119 @@
+# Building Good Containers
+
+There are a number of recommendations based on our experience building
+Node.js applications in containers for deployments:
+
+* build non-root containers
+* avoid using trusted ports
+* use multi-stage builds
+* add key tools for problem determination
+* use a process manager 
+
+## Build non-root containers
+
+Running processes as root, even in containers can be a security risk,
+particularly if external resources are mapped into the container. 
+The following is a good article which explains why:
+[Processes In Containers Should Not Run As Root](https://medium.com/@mccode/processes-in-containers-should-not-run-as-root-2feae3f0df3b#:~:text=Containers%20are%20not%20trust%20boundaries,a%20container%20on%20your%20server)
+
+There are also systems which support running containers as `rootless`.
+In order to ensure that your containers can run on these system you
+want to avoid running process as root as they may end up depending
+on capabilities that are not available when they are run `rootless`.
+
+When using docker/podman you can change the user with the `USER` command.
+Well designed base containers will have already done this, but you should
+always check what user will be used by default.
+
+```shell
+USER 1001
+RUN chown -R 1001:0 /some/directory
+```
+
+It is recommended that you build your containers as `non-root`.
```

Comment:
What is meant by `building` the container? For me the `docker build` command is used to build a container. Do we instead want to say `run processes as non-root inside the container`? or something similar.

Reply:
+1